### PR TITLE
ZEPPELIN-371 : Import a notebook

### PR DESCRIPTION
**Pull request by prabhjyotsingh**
Import a notebook by either an external URL or uploading a valid JSON.

screen shot 2015-10-28 at 4 53 58 pm

screen shot 2015-10-28 at 4 56 30 pm

screen shot 2015-10-28 at 4 56 55 pm

screen shot 2015-10-28 at 4 55 29 pm

### CODE REVIEW

**Comment by felixcheung** [comment]
could we add test for this?

**Comment by prabhjyotsingh** [comment]
Sure, I'll write a unit test case for this.

**Comment by prabhjyotsingh** [comment]
Added test case for the same as well.

CI fails in this PR. with following reason

    [32;1mThe command "mvn package -Pbuild-distr -Phadoop-2.3 -Ppyspark -B" exited with 0.[0m
    travis_time:start:088fd37b
    [0K$ ./testing/startSparkCluster.sh 1.4.0 2.3
    SPARK_HOME is /home/travis/build/apache/incubator-zeppelin/spark-1.4.0-bin-hadoop2.3 
    
    
    No output has been received in the last 10 minutes, this potentially indicates a stalled build or something wrong with the build itself.

    The build has been terminated

**Comment by Leemoonsoo** [comment]
Thanks @prabhjyotsingh for the nice feature. I manually tested and working well for me.
CI failing is fixed by #380. Could you try rebase this branch?

**Comment by prabhjyotsingh** [comment]
The build is failing with following error

        06:21:48,776  INFO org.apache.zeppelin.rest.AbstractTestRestApi:226 - Test Zeppelin terminated.
    Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 4.115 sec <<< FAILURE! - in org.apache.zeppelin.rest.ZeppelinRestApiTest
    org.apache.zeppelin.rest.ZeppelinRestApiTest  Time elapsed: 4.115 sec  <<< ERROR!
    java.lang.NullPointerException: null
    at org.apache.zeppelin.rest.AbstractTestRestApi.startUp(AbstractTestRestApi.java:123)
    at org.apache.zeppelin.rest.ZeppelinRestApiTest.init(ZeppelinRestApiTest.java:57)

    Running org.apache.zeppelin.security.SecurityUtilsTest
    Tests run: 9, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.032 sec - in org.apache.zeppelin.security.SecurityUtilsTest
    Running org.apache.zeppelin.server.CorsFilterTest
    Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.29 sec - in org.apache.zeppelin.server.CorsFilterTest
    06:21:49,112  INFO org.apache.zeppelin.server.ZeppelinServer:112 - Shutting down Zeppelin Server ...
    06:21:49,112  INFO org.apache.zeppelin.server.ZeppelinServer:112 - Shutting down Zeppelin Server ...
    06:21:49,113  INFO org.apache.zeppelin.server.ZeppelinServer:118 - Bye
    06:21:49,113  INFO org.apache.zeppelin.server.ZeppelinServer:118 - Bye

    Results :

    Tests in error:
    ZeppelinSparkClusterTest.init:46->AbstractTestRestApi.startUp:123 Â» NullPointer
    ZeppelinRestApiTest.init:57->AbstractTestRestApi.startUp:123 Â» NullPointer

    Tests run: 16, Failures: 0, Errors: 2, Skipped: 0

    [INFO] ------------------------------------------------------------------------
    [INFO] Reactor Summary:
    [INFO]
    [INFO] Zeppelin ........................................... SUCCESS [ 10.558 s]
    [INFO] Zeppelin: Interpreter .............................. SUCCESS [01:19 min]
    [INFO] Zeppelin: Zengine .................................. SUCCESS [  4.987 s]
    [INFO] Zeppelin: Spark dependencies ....................... SUCCESS [ 32.026 s]
    [INFO] Zeppelin: Spark .................................... SUCCESS [01:20 min]
    [INFO] Zeppelin: Markdown interpreter ..................... SUCCESS [  0.700 s]
    [INFO] Zeppelin: Angular interpreter ...................... SUCCESS [  0.279 s]
    [INFO] Zeppelin: Shell interpreter ........................ SUCCESS [  0.356 s]
    [INFO] Zeppelin: Hive interpreter ......................... SUCCESS [  2.433 s]
    [INFO] Zeppelin: Apache Phoenix Interpreter ............... SUCCESS [ 10.156 s]
    [INFO] Zeppelin: PostgreSQL interpreter ................... SUCCESS [  1.127 s]
    [INFO] Zeppelin: Tajo interpreter ......................... SUCCESS [  1.565 s]
    [INFO] Zeppelin: Flink .................................... SUCCESS [ 18.759 s]
    [INFO] Zeppelin: Apache Ignite interpreter ................ SUCCESS [ 15.523 s]
    [INFO] Zeppelin: Kylin interpreter ........................ SUCCESS [  0.678 s]
    [INFO] Zeppelin: Lens interpreter ......................... SUCCESS [  3.232 s]
    [INFO] Zeppelin: Cassandra ................................ SUCCESS [ 39.078 s]
    [INFO] Zeppelin: web Application .......................... SUCCESS [ 37.655 s]
    [INFO] Zeppelin: Server ................................... FAILURE [ 15.121 s]
    [INFO] Zeppelin: Packaging distribution ................... SKIPPED
    [INFO] ------------------------------------------------------------------------
    [INFO] BUILD FAILURE
    [INFO] ------------------------------------------------------------------------
    [INFO] Total time: 05:55 min
    [INFO] Finished at: 2015-11-06T06:21:49+00:00
    [INFO] Final Memory: 66M/362M
    [INFO] ------------------------------------------------------------------------
    [ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.17:test (default-test) on project zeppelin-server: There are test failures.
    [ERROR]
    [ERROR] Please refer to /home/travis/build/apache/incubator-zeppelin/zeppelin-server/target/surefire-reports for the individual test results.
    [ERROR] -> [Help 1]
    [ERROR]
    [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
    [ERROR] Re-run Maven using the -X switch to enable full debug logging.
    [ERROR]
    [ERROR] For more information about the errors and possible solutions, please read the following articles:
    [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
    [ERROR]
    [ERROR] After correcting the problems, you can resume the build with the command
    [ERROR]   mvn <goals> -rf :zeppelin-server
    travis_time:end:1b88a8e0:start=1446790550157092954,finish=1446790909997950807,duration=359840857853
    [0K
    [31;1mThe command "mvn verify -Pusing-packaged-distr -Phadoop-2.3 -Ppyspark -B" exited with 1.[0m



My guess is, in AbstractTestRestApi:116 
ZeppelinServer.notebook.getInterpreterFactory().get() is empty
any pointers what would have gone wrong ?

**Comment by Leemoonsoo** [comment]
I'm in the same CI test error in #341. Let me try to figure out.

**Comment by Leemoonsoo** [comment]
Please take a look [NotebookServerTest](https://github.com/apache/incubator-zeppelin/blob/master/zeppelin-server/src/test/java/org/apache/zeppelin/socket/NotebookServerTest.java). Extending AbstractTestRestApi helped me to make ci build passes in #341.

**Comment by prabhjyotsingh** [comment]
Thanks @Leemoonsoo, have made relevant changes.

now it fails for  "ZeppelinIT.testAngularDisplay:173->createNewNote:306 » Timeout Timed out after"

    -------------------------------------------------------
    T E S T S
    -------------------------------------------------------
    Running org.apache.zeppelin.ZeppelinIT
    Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 22.224 sec <<< FAILURE! - in org.apache.zeppelin.ZeppelinIT
    testAngularDisplay(org.apache.zeppelin.ZeppelinIT)  Time elapsed: 22.152 sec  <<< ERROR!
    org.openqa.selenium.TimeoutException: Timed out after 10 seconds waiting for visibility of element located by By.id: noteNameModal
    Build info: version: '2.44.0', revision: '76d78cf323ce037c5f92db6c1bba601c2ac43ad8', time: '2014-10-23 13:11:40'
    System info: host: 'testing-worker-linux-docker-095ed8f0-3208-linux-3', ip: '172.17.12.223', os.name: 'Linux', os.arch: 'amd64', os.version: '3.13.0-40-generic', java.version: '1.7.0_76'
    Session ID: e6ed0c8a-b983-40da-b950-09ead342ffe0
    Driver info: org.openqa.selenium.firefox.FirefoxDriver
    Capabilities [{platform=LINUX, acceptSslCerts=true, javascriptEnabled=true, cssSelectorsEnabled=true, databaseEnabled=true, browserName=firefox, handlesAlerts=true, nativeEvents=false, webStorageEnabled=true, rotatable=false, locationContextEnabled=true, applicationCacheEnabled=true, takesScreenshot=true, version=31.0}]
        at org.openqa.selenium.support.ui.WebDriverWait.timeoutException(WebDriverWait.java:79)
        at org.openqa.selenium.support.ui.FluentWait.until(FluentWait.java:228)
        at org.apache.zeppelin.ZeppelinIT.createNewNote(ZeppelinIT.java:306)
        at org.apache.zeppelin.ZeppelinIT.testAngularDisplay(ZeppelinIT.java:173)


    Results :

    Tests in error:
      ZeppelinIT.testAngularDisplay:173->createNewNote:306 » Timeout Timed out after...

    Tests run: 1, Failures: 0, Errors: 1, Skipped: 0




Trying to debug, will update once finished.

File: zeppelin-web/src/index.html [hunk 1]
```diff
@@ -45,6 +45,7 @@
     <link rel="stylesheet" href="bower_components/ngtoast/dist/ngToast.css" />
     <!-- endbower -->
     <link rel="stylesheet" href="bower_components/jquery-ui/themes/base/all.css" />
+    <link rel="stylesheet" href="bower_components/ngAnimate/css/ng-animation.css" />
```
Comment (Leemoonsoo) [inline zeppelin-web/src/index.html]: Could you add license of ng-animation.css in zeppelin-distribution/src/bin_license/LICENSE file? 
For example ngToast, https://github.com/apache/incubator-zeppelin/blob/master/zeppelin-distribution/src/bin_license/LICENSE#L95

**Comment by prabhjyotsingh** [comment]
Thanks @Leemoonsoo while importing any external library I'll keep this in mind for next time.

Resolved build failure, ready for review.

**Comment by Leemoonsoo** [comment]
Working nicely, Looks good to me.

**Comment by Leemoonsoo** [comment]
Merge if there're no more discussions

File: zeppelin-server/src/main/java/org/apache/zeppelin/socket/NotebookServer.java [hunk 2]
```diff
@@ -152,6 +151,9 @@ public void onMessage(NotebookSocket conn, String msg) {
           case ANGULAR_OBJECT_UPDATED:
             angularObjectUpdated(conn, notebook, messagereceived);
             break;
+          case IMPORT_NOTEBOOK:
```
Comment (mindprince) [inline zeppelin-server/src/main/java/org/apache/zeppelin/socket/NotebookServer.java]: Can we have uniform naming? All the other operations are called `XYZ_NOTE`. Let's follow that here as well.

Comment (prabhjyotsingh) [inline zeppelin-server/src/main/java/org/apache/zeppelin/socket/NotebookServer.java]: Sure @mindprince, that make sense, did that. Thanks.

**Comment by bzz** [comment]
@prabhjyotsingh thank you, great feature. 
Huge :+1: for having a tests!

This comes very handy with services like github or https://www.zeppelinhub.com/viewer

**Pull request closed without merging**